flask/app.py

```diff
@@ -899,12 +899,28 @@ def run(self, host=None, port=None, debug=None,
             The default port is now picked from the ``SERVER_NAME``
             variable.
         """
+        # determine if this method is being invoked in the standard way (from
+        # the global scope of the __main__ module) by checking the length of
+        # the call stack
+        from inspect import stack
+        called_from_main = len(stack()) == 2
+
         # Change this into a no-op if the server is invoked from the
         # command line. Have a look at cli.py for more information.
-        if os.environ.get('FLASK_RUN_FROM_CLI') == 'true':
-            from .debughelpers import explain_ignored_app_run
-            explain_ignored_app_run()
-            return
+        if called_from_main:
+            if os.environ.get('FLASK_RUN_FROM_CLI') == 'true':
+                if os.environ.get('FLASK_RUN_FROM_APP_RUN') != 'true':
+                    from .debughelpers import explain_ignored_app_run
+                    explain_ignored_app_run()
+                return
+
+            if load_dotenv and not options:
```

Comment:
I'm uncomfortable with this part long term. I think people are going to be surprised when suddenly their `python myapp.py shell` command fails because they passed another option to `app.run`.